### PR TITLE
[NTOS:PS] Implement PS_QUOTA_TYPE and let the quota code use it

### PR DESCRIPTION
**Pull request by GeoB99**
Pretty much what the PR title and commit description say. Tested the patch on my backend, ROS works quite nicely without nasty issues or the like.

### CODE REVIEW

File: ntoskrnl/ps/quota.c [hunk 1]
```diff
@@ -28,66 +28,62 @@ static KSPIN_LOCK PspQuotaLock;
...
 NTAPI
 PspChargeProcessQuotaSpecifiedPool(IN PEPROCESS Process,
-                                   IN UCHAR     PoolIndex,
+                                   IN PS_QUOTA_TYPE QuotaType,
                                    IN SIZE_T    Amount)
 {
     ASSERT(Process);
     ASSERT(Process != PsInitialSystemProcess);
-    ASSERT(PoolIndex <= 2);
+    ASSERT(QuotaType <= 2);
```
Comment (binarymaster) [inline ntoskrnl/ps/quota.c:43]:
```suggestion
    ASSERT(QuotaType < PsQuotaTypes);
```

File: ntoskrnl/ps/quota.c [hunk 2]
```diff
@@ -28,66 +28,62 @@ static KSPIN_LOCK PspQuotaLock;
...
 NTAPI
 PspReturnProcessQuotaSpecifiedPool(IN PEPROCESS Process,
-                                   IN UCHAR     PoolIndex,
+                                   IN PS_QUOTA_TYPE QuotaType,
                                    IN SIZE_T    Amount)
 {
     ASSERT(Process);
     ASSERT(Process != PsInitialSystemProcess);
-    ASSERT(PoolIndex <= 2);
+    ASSERT(QuotaType <= 2);
```
Comment (binarymaster) [inline ntoskrnl/ps/quota.c:78]:
```suggestion
    ASSERT(QuotaType < PsQuotaTypes);
```

**Review by binarymaster — COMMENTED** [review]
Since you're getting rid of magic values here, let's do it completely. :slightly_smiling_face:

File: sdk/include/ndk/pstypes.h [hunk 3]
```diff
@@ -991,6 +991,17 @@ typedef struct _EPROCESS_QUOTA_BLOCK
...
+//
+// Quota Structures
+//
+typedef enum _PS_QUOTA_TYPE
+{
+    PsNonPagedPool,
+    PsPagedPool,
+    PsPageFile,
+    PsQuotaTypes
+} PS_QUOTA_TYPE;
```
Comment (SergeGautherie) [inline sdk/include/ndk/pstypes.h:1003]: See https://www.geoffchappell.com/studies/windows/km/ntoskrnl/api/ps/psquota/type.htm
You may want to support the two other values too... (At least, a comment?)

File: sdk/include/ndk/pstypes.h [hunk 4]
```diff
@@ -991,6 +991,17 @@ typedef struct _EPROCESS_QUOTA_BLOCK
     ULONG ProcessCount;
 } EPROCESS_QUOTA_BLOCK, *PEPROCESS_QUOTA_BLOCK;
 
+//
+// Quota Structures
+//
+typedef enum _PS_QUOTA_TYPE
+{
+    PsNonPagedPool,
```
Comment (HBelusca) [inline sdk/include/ndk/pstypes.h:999]:
```suggestion
    PsNonPagedPool = 0,
```

File: sdk/include/ndk/pstypes.h [hunk 5]
```diff
@@ -991,6 +991,26 @@ typedef struct _EPROCESS_QUOTA_BLOCK
...
+#if (NTDDI_VERSION >= NTDDI_LONGHORN)
+    PsWorkingSet,
+    PsQuotaTypes,
+#elif (NTDDI_VERSION == NTDDI_LONGHORN)
+    PsWorkingSet,
+    PsCpuRate,
+    PsQuotaTypes,
+#else
+    PsQuotaTypes
+#endif
```
Comment (binarymaster) [inline sdk/include/ndk/pstypes.h:1011]: Too many `PsQuotaTypes` :)
```suggestion
#if (NTDDI_VERSION >= NTDDI_LONGHORN)
    PsWorkingSet,
#elif (NTDDI_VERSION == NTDDI_LONGHORN)
    PsWorkingSet,
    PsCpuRate,
#endif
    PsQuotaTypes
```

File: sdk/include/ndk/pstypes.h [hunk 6]
```diff
@@ -991,6 +991,26 @@ typedef struct _EPROCESS_QUOTA_BLOCK
     ULONG ProcessCount;
 } EPROCESS_QUOTA_BLOCK, *PEPROCESS_QUOTA_BLOCK;
 
+//
+// Quota Structures
```
Comment (SergeGautherie) [inline sdk/include/ndk/pstypes.h:995]: There is no `struct` here...

**Review by binarymaster — APPROVED** [review]

**Comment by SergeGautherie** [comment]
> let's do it completely

There seems to be even more cases: for example,
https://git.reactos.org/?p=reactos.git&a=search&h=HEAD&st=grep&s=QuotaUsage

And note that `/* 0=PagedPool, 1=NonPagedPool, 2=Pagefile */` does not match this PR, so should be double-checked.

**Comment by Fraizeraust** [comment]
> And note that `/* 0=PagedPool, 1=NonPagedPool, 2=Pagefile */` does not match this PR, so should be double-checked.

According to what is being told in the documentation by Geoff regarding `EPROCESS` structure -- https://geoffchappell.com/studies/windows/km/ntoskrnl/inc/ps/eprocess/index.htm

> Version 5.1 brought the pools and pagefile together into a PS_QUOTA_TYPE enumeration that indexes the QuotaUsage and QuotaPeak arrays, which thus absorb what had been PagefileUsage and PeakPagefileUsage. 


And from the `PsChargePoolQuota` function -- https://www.geoffchappell.com/studies/windows/km/ntoskrnl/api/ps/psquota/chargepoolquota.htm?tx=68,142

> Versions 5.1 and 5.2 interpret PoolType as if from the PS_QUOTA_TYPE enumeration and simply assume that it is one of PsNonPagedPool (0), PsPagedPool (1) or PsPageFile (2).

**POOL_TYPE**, in other words, implicitly depends upon the PS_QUOTA_TYPE enumeration in the aforementioned function. As a quick research, I put a breakpoint on that kernel routine to investigate in which circumstances the function gets called.
Capture
The **1** number pushed into the stack is the second parameter, the PoolType. The function is called in `nt!FsRtlNotifyFilterReportChange+0x4c2`, specifically when you launch Disk Cleanup to begin scanning the files in order to hit the breakpoint. Certainly such scenario can be described as the pool memory being paged, not otherwise. I highly doubt the pool is non-paged. Whoever that person who wrote the comment got `PagedPool` and `NonPagedPool` values inverted.

**Comment by HBelusca** [comment]
> > And note that `/* 0=PagedPool, 1=NonPagedPool, 2=Pagefile */` does not match this PR, so should be double-checked.
>
> Certainly such scenario can be described as the pool memory being paged, not otherwise. I highly doubt the pool is non-paged. Whoever that person wrote the comment got `PagedPool` and `NonPagedPool` values inverted.

Nevertheless these values (0 and 1) are used in consistence with that comment (and not with the PS_QUOTA_TYPE) when you see from the grep results, that the values obtained from the arrays are assigned to variables having similar names.
For example:
```
1026                 SpiCurrent->QuotaPagedPoolUsage = Process->QuotaUsage[0];
1028                 SpiCurrent->QuotaNonPagedPoolUsage = Process->QuotaUsage[1];
```
and the usages of the QuotaUsage arrays in comparisons in functions....

File: sdk/include/ndk/pstypes.h [hunk 7]
```diff
@@ -1208,7 +1229,7 @@ typedef struct _EPROCESS
     EX_RUNDOWN_REF RundownProtect;
     HANDLE UniqueProcessId;
     LIST_ENTRY ActiveProcessLinks;
-    SIZE_T QuotaUsage[3]; /* 0=PagedPool, 1=NonPagedPool, 2=Pagefile */
+    SIZE_T QuotaUsage[3]; /* 0=NonPagedPool, 1=PagedPool, 2=Pagefile */
```
Comment (SergeGautherie) [inline sdk/include/ndk/pstypes.h:1232]:
```suggestion
    SIZE_T QuotaUsage[3]; // See PS_QUOTA_TYPE.
```

**Review by SergeGautherie — CHANGES_REQUESTED** [review]

File: sdk/include/ndk/pstypes.h [hunk 8]
```diff
@@ -1208,7 +1229,7 @@ typedef struct _EPROCESS
...
     SIZE_T QuotaPeak[3];  /* ditto */
```
Comment (binarymaster) [inline sdk/include/ndk/pstypes.h:1233]:
```suggestion
    SIZE_T QuotaPeak[PsQuotaTypes];  /* ditto */
```

File: sdk/include/ndk/pstypes.h [hunk 9]
```diff
@@ -985,7 +1002,11 @@ typedef struct _EPROCESS_QUOTA_ENTRY
 
 typedef struct _EPROCESS_QUOTA_BLOCK
 {
-    EPROCESS_QUOTA_ENTRY QuotaEntry[3];
+#if (NTDDI_VERSION == NTDDI_LONGHORN)
+    EPROCESS_QUOTA_ENTRY QuotaEntry[PsQuotaTypes - 1];
+#else
+    EPROCESS_QUOTA_ENTRY QuotaEntry[PsQuotaTypes];
+#endif
```
Comment (binarymaster) [inline sdk/include/ndk/pstypes.h:1009]: Why this check here? `PsQuotaTypes` would already have different value based on the enum definition.
```suggestion
    EPROCESS_QUOTA_ENTRY QuotaEntry[PsQuotaTypes];
```

File: ntoskrnl/ps/quota.c [hunk 10]
```diff
@@ -28,66 +28,62 @@ static KSPIN_LOCK PspQuotaLock;
...
 NTAPI
 PspChargeProcessQuotaSpecifiedPool(IN PEPROCESS Process,
-                                   IN UCHAR     PoolIndex,
+                                   IN PS_QUOTA_TYPE QuotaType,
                                    IN SIZE_T    Amount)
 {
     ASSERT(Process);
     ASSERT(Process != PsInitialSystemProcess);
-    ASSERT(PoolIndex <= 2);
+    ASSERT(QuotaType < PsQuotaTypes);
```
Comment (SergeGautherie) [inline ntoskrnl/ps/quota.c:43]: NT 5.1/5.2:
```suggestion
    ASSERT(QuotaType < 3); // See EPROCESS.
```

NT 6.0:
postponed.

File: sdk/include/ndk/pstypes.h (hunk 7 above)
Comment (SergeGautherie) [reply sdk/include/ndk/pstypes.h:1232]: @binarymaster, Note that I do not concur to using `PsQuotaTypes` here.

File: ntoskrnl/ps/quota.c (hunk 10 above)
Comment (Fraizeraust) [reply ntoskrnl/ps/quota.c:43]: lol

With your comment you're telling me to go back using magic values, that's a big no.

Comment (SergeGautherie) [reply ntoskrnl/ps/quota.c:43]: Not really: I am merely suggesting to be consistent with geoffchappell documentation (here and elsewhere), unless the latter is proven wrong...

File: sdk/include/ndk/pstypes.h [hunk 11]
```diff
@@ -1208,8 +1225,8 @@ typedef struct _EPROCESS
     EX_RUNDOWN_REF RundownProtect;
     HANDLE UniqueProcessId;
     LIST_ENTRY ActiveProcessLinks;
-    SIZE_T QuotaUsage[3]; /* 0=PagedPool, 1=NonPagedPool, 2=Pagefile */
-    SIZE_T QuotaPeak[3];  /* ditto */
+    SIZE_T QuotaUsage[PsQuotaTypes]; /* See PS_QUOTA_TYPE */
+    SIZE_T QuotaPeak[PsQuotaTypes];  /* ditto */
```
Comment (SergeGautherie) [inline sdk/include/ndk/pstypes.h:1229]: There is no need anymore for these comments, if you (ab)use `PsQuotaTypes`.

**Review by encodedpr — APPROVED** [review]

File: sdk/include/ndk/pstypes.h [hunk 12]
```diff
@@ -972,6 +972,23 @@ typedef struct _JOB_SET_ARRAY
...
+//
+typedef enum _PS_QUOTA_TYPE
+{
+    PsNonPagedPool = 0,
+    PsPagedPool,
+    PsPageFile,
+#if (NTDDI_VERSION >= NTDDI_LONGHORN)
+    PsWorkingSet,
+#endif
+#if (NTDDI_VERSION == NTDDI_LONGHORN)
```
Comment (Extravert-ir) [inline sdk/include/ndk/pstypes.h:986]: What's the logic here? `PsCpuRate` was removed after Vista?

Comment (HBelusca) [reply sdk/include/ndk/pstypes.h:986]: According to Geoff Chappell https://www.geoffchappell.com/studies/windows/km/ntoskrnl/api/ps/psquota/type.htm :
```
Value | Name | Versions
  --  |  --  |   --
  0  | PsNonPagedPool | 5.1 and higher
  1  | PsPagedPool    | 5.1 and higher
  2  | PsPageFile     | 5.1 and higher
  3  | working set pages | 6.0 and higher
**4  | CPU rate limit    | 6.0 to 6.2 ** <--- This one
  3 (5.1 to 5.2);
  5 (6.0 to 6.2);  | PsQuotaTypes | 5.1 and higher
  4
```

**Pull request merged**